File: ice-core/src/main/java/com/ice/core/relation/None.java
package com.ice.core.relation;

import com.ice.common.enums.NodeRunStateEnum;
import com.ice.core.base.BaseNode;
import com.ice.core.base.BaseRelation;
import com.ice.core.context.IceContext;
import com.ice.core.utils.IceLinkedList;

/**
 * @author waitmoon
 * relation NONE
 * all child will execute
 * return NONE
 */
public final class None extends BaseRelation {
    /*
     * process relation none
     */
    @Override
    protected NodeRunStateEnum processNode(IceContext ctx) {
        IceLinkedList<BaseNode> children = this.getChildren();
        if (children == null || children.isEmpty()) {
            return NodeRunStateEnum.NONE;
        }
        if (children.getSize() == 1) {
            BaseNode node = children.get(0);
            if (node == null) {
                return NodeRunStateEnum.NONE;
            }
            return node.process(ctx);
        }
        for (IceLinkedList.Node<BaseNode> listNode = children.getFirst(); listNode != null; listNode = listNode.next) {
            BaseNode node = listNode.item;
            if (node != null) {
                node.process(ctx);
            }
        }

        return NodeRunStateEnum.NONE;
    }
}

File: sdks/go/relation/none.go
package relation

import (
	stdctx "context"

	icecontext "github.com/zjn-zjn/ice/sdks/go/context"
	"github.com/zjn-zjn/ice/sdks/go/enum"
	"github.com/zjn-zjn/ice/sdks/go/node"
)

// None is a relation node that executes all children and always returns NONE.
type None struct {
	node.Relation
}

// NewNone creates a new None relation node.
func NewNone() *None {
	return &None{
		Relation: *node.NewRelation(),
	}
}

// Process implements the Node interface.
func (n *None) Process(ctx stdctx.Context, roam *icecontext.Roam) enum.RunState {
	return node.ProcessWithBase(ctx, &n.Base, roam, n.processNode, nil)
}

func (n *None) processNode(ctx stdctx.Context, roam *icecontext.Roam) enum.RunState {
	if n.Children == nil || n.Children.IsEmpty() {
		return enum.NONE
	}

	for listNode := n.Children.First(); listNode != nil; listNode = listNode.Next {
		child := listNode.Item
		if child != nil {
			child.Process(ctx, roam)
		}
	}

	return enum.NONE
}
